specs/opcode/5AGAS.md

```diff
@@ -0,0 +1,40 @@
+# GAS opcode
+
+## Procedure
+
+The `GAS` opcode gets the amount of available gas, including the correspoding reduction for the cost of the instruction itself.
+
+## EVM behaviour
+
+The `GAS` opcode loads a 64-bit value for the available gas from the transaction context, and pushes it to the stack.
+
+## Circuit behaviour
+
+1. construct tx table
+2. bus-mapping lookup for stack write operation
+3. call context lookup (for `tx_id`)
+4. tx lookup (for `gas_left`)
```

Comment:
I'm not really sure if the tx table contains a `gas_left` field.  As far as I know, the Tx Table only contains static information about each Tx which correspond 1 to 1 with the block txs (which are public inputs).  So I think that `TxContextFieldTag.Gas` contains the gas that the gas amount set for that transaction.

I didn't see any tag in any table used to track the `gas_left` at the step granularity.  Maybe we need to introduce it? (In the `rw_table` I'd say)

Or maybe I got confused in my understanding of the Tx Table?  I hope @han0110 can help us here!

Reply:
Hi @ed255 thanks for pointing this out.

I think I did not update the specs here (after having modified the circuits implementation).

As far as the circuits impl goes, we compare against the gas left in the contraint builder's current state (subtracted by the gas cost for the GAS opcode itself): [reference](https://github.com/scroll-tech/zkevm-circuits/blob/feat/gas-gadget/zkevm-circuits/src/evm_circuit/execution/gas.rs#L36-L43).

This was after realisation that the `tx lookup` table is not responsible for storing the gas field.

I will update the spec PR.

Reply:
> I'm not really sure if the tx table contains a `gas_left` field. As far as I know, the Tx Table only contains static information about each Tx which correspond 1 to 1 with the block txs (which are public inputs). So I think that `TxContextFieldTag.Gas` contains the gas that the gas amount set for that transaction.
> 
> I didn't see any tag in any table used to track the `gas_left` at the step granularity. Maybe we need to introduce it? (In the `rw_table` I'd say)
> 
> Or maybe I got confused in my understanding of the Tx Table? I hope @han0110 can help us here!
> 
> EDIT: I've set my review as `requested changes` just to avoid merging this until the doubt is resolved (I'm not actually requesting any change right now) :)

Thanks, @ed255 ! Sorry for missing that out.

Reply:
I guess we will need to pass it from the bus-mapping and annotate it in a table, I agree. Otherways the data used here simply is not avaliable at circuit/proof setup time.

Good catch @ed255!

Should we move into doing a PR for this? Also, which are your thoughts @han0110 ?

By looking at https://github.com/appliedzkp/zkevm-circuits/pull/301/ in the impl. They simply get the gas_left from the `ConstraintBuilder::curr` . Which basically (IIUC) does not match what is mentioned here as no lookups are performed there.

Reply:
> Should we move into doing a PR for this? Also, which are your thoughts @han0110 ?
> 
> By looking at [appliedzkp/zkevm-circuits#301](https://github.com/appliedzkp/zkevm-circuits/pull/301) in the impl. They simply get the gas_left from the `ConstraintBuilder::curr` . Which basically (IIUC) does not match what is mentioned here as no lookups are performed there.

I would say that taking the gas from the constraint builder's current state is a good solution so there will be no need to add the gas_left to a table as I originally thought.  And the spec has now been updated to reflect that.

Reply:
Makes sense. 

Thought that the lookup was needed. But after seeing that was an error and the refactor, that looks like a much better solution!!!